simulator/main.rs

```diff
@@ -67,6 +113,317 @@ enum Value {
+                let mut t = t.clone();
+                if rng.gen_bool(0.01) {
+                    t.pop();
+                    LTValue(Value::Text(t))
+                } else {
+                    let index = rng.gen_range(0..t.len());
+                    let mut t = t.into_bytes();
+                    t[index] -= 1;
+                    // Mutate the rest of the string
+                    for i in (index + 1)..t.len() {
+                        t[i] = rng.gen_range(0..=255);
+                    }
+                    LTValue(Value::Text(String::from_utf8(t).unwrap()))
+                }
+            }
+            Value::Blob(b) => todo!(),
+            _ => unreachable!(),
+        }
+    }
+}
+
+struct GTValue(Value);
+
+impl ArbitraryOf<Vec<&Value>> for GTValue {
+    fn arbitrary_of<R: Rng>(rng: &mut R, t: &Vec<&Value>) -> Self {
+        if t.is_empty() {
+            return GTValue(Value::Null);
+        }
+
+        let index = rng.gen_range(0..t.len());
+        GTValue::arbitrary_of(rng, t[index])
+    }
+}
+
+impl ArbitraryOf<Value> for GTValue {
+    fn arbitrary_of<R: Rng>(rng: &mut R, t: &Value) -> Self {
+        match t {
+            Value::Integer(i) => GTValue(Value::Integer(rng.gen_range(*i..i64::MAX))),
+            Value::Float(f) => GTValue(Value::Float(rng.gen_range(*f..1e10))),
+            Value::Text(t) => {
+                // Either lengthen the string, or make at least one character smaller and mutate the rest
+                let mut t = t.clone();
+                if rng.gen_bool(0.01) {
+                    t.push(rng.gen_range(0..=255) as u8 as char);
+                    GTValue(Value::Text(t))
+                } else {
+                    let index = rng.gen_range(0..t.len());
+                    let mut t = t.into_bytes();
+                    t[index] += 1;
+                    // Mutate the rest of the string
+                    for i in (index + 1)..t.len() {
+                        t[i] = rng.gen_range(0..=255);
+                    }
+                    GTValue(Value::Text(String::from_utf8(t).unwrap()))
+                }
+            }
+            Value::Blob(b) => todo!(),
+            _ => unreachable!(),
+        }
+    }
+}
+
+enum Predicate {
+    And(Vec<Predicate>),
+    Or(Vec<Predicate>),
+    Eq(String, Value),
+    Gt(String, Value),
+    Lt(String, Value),
+}
+
+enum Query {
+    Create { table: Table },
+    Select { table: String, guard: Predicate },
+    Insert { table: String, values: Vec<Value> },
+    Delete { table: String, guard: Predicate },
+}
+
+impl ArbitraryOf<Table> for Query {
+    fn arbitrary_of<R: Rng>(rng: &mut R, t: &Table) -> Self {
+        match rng.gen_range(0..=200) {
+            0 => Query::Create {
+                table: Table::arbitrary(rng),
+            },
+            1..=100 => Query::Select {
+                table: t.name.clone(),
+                guard: Predicate::arbitrary_of(rng, t),
```

Comment:
IMO either call this `where` or `predicate`, since we already use a `Predicate` type here (and the `predicate` name elsewhere in Limbo too, actually)

Reply:
We can't call it where, as Rust already has `where` as a keyword. I changed it to `predicate`.